pkg/pillar/zedUpload/awsutil/s3sync.go

```diff
@@ -101,6 +111,21 @@ func (r *CustomWriter) Seek(offset int64, whence int) (int64, error) {
 	return r.fp.Seek(offset, whence)
 }
 
+// ChunkWriter encapsulates the chunk data and the channel to write it back to caller
+type ChunkWriter struct {
+	data      ChunkStats
+	prgNotify ChunkStatsChan
+}
+
+// WriteAt is the overridden method of io.WriterAt method, which helps in chunk by chunk downloading.
+func (r *ChunkWriter) WriteAt(p []byte, off int64) (int, error) {
+	if r.prgNotify != nil {
+		r.data.Chunk = p
```

Comment:
Looks like nothing is reading this. Correct?

Reply:
> The time.Sleep really concerns me. Please explain.
> 
> Do we need to support SysOpDownloadByChunks for other datastores like http (with an implementation which doesn't chunk if the protocol doesn't support chunks)? If not, how does the users of this package know which datastores which methods?
> 
> We should have unittests for this but the tests can not contain credentials.
> 
> Names are confusing.
> 
> And is the code duplication necessary? Seems like the struct definiting and Writer are identical and can be shared.


time.Sleep -- > as and when the small chunk of data is download, I was writing into the prgChan channel and in-turn prgChan channel was post the data in req.result channel, the method was returning the moment we write on the prgChan, irrespective of whether channel finishes it's work.

SysOpDownloadByChunks --> supported for s3 and azure not sure about the other datastore

test case --> will running as part of zedCloud apiTest

Naming confusion --> fixing it

Duplicate code --> I didn't wanted to touch currently existing code/ struct, will fix the duplicate code issue.

Reply:
sorry! I didn't get you

Reply:
What is reading data.Chunk? (I think it is in completely different part of the code, with the fundamental confusion being between the names - think is not a notification channel it is a data+notification channel AFAICT)

Reply:
Yes, as you said, it's data+ notification, upon writing into the channel, the values on the channel is read and the data is passed the caller.